haystack/preview/components/preprocessors/text_document_splitter.py

```diff
@@ -0,0 +1,96 @@
+        "sentence" for splitting by ".", or "passage" for splitting by "\n\n".
+        :param split_length: The maximum number of units in each split.
+        :param split_overlap: The number of units that each split should overlap.
+        """
+
+        self.split_by = split_by
+        if split_by not in ["word", "sentence", "passage"]:
+            raise ValueError("split_by must be one of 'word', 'sentence' or 'passage'.")
+        if split_length <= 0:
+            raise ValueError("split_length must be greater than 0.")
+        self.split_length = split_length
+        if split_overlap < 0:
+            raise ValueError("split_overlap must be greater than or equal to 0.")
+        self.split_overlap = split_overlap
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document]):
+        """
+        Splits the documents by split_by after split_length units with an overlap of split_overlap units.
+        Returns a list of documents with the split texts.
+        :param documents: The documents to split.
+        :return: A list of documents with the split texts.
+        """
+
+        if not documents or not isinstance(documents, list) or not isinstance(documents[0], Document):
+            raise TypeError("TextDocumentSplitter expects a List of Documents as input.")
+        split_docs = []
+        for doc in documents:
+            if doc.text is None:
+                raise ValueError("TextDocumentSplitter only works with text documents but document.text is None.")
+            units, split_at = self._split_into_units(doc.text, self.split_by)
+            text_splits = self._concatenate_units(units, self.split_length, self.split_overlap, split_at)
+            metadata = doc.metadata
+            metadata["source_id"] = doc.id
+            split_docs += [Document(text=txt, metadata=metadata) for txt in text_splits]
+        return {"documents": split_docs}
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self, split_by=self.split_by, split_length=self.split_length, split_overlap=self.split_overlap
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "TextDocumentSplitter":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    def _split_into_units(self, text: str, split_by: Literal["word", "sentence", "passage"]) -> Tuple[List[str], str]:
+        if split_by == "passage":
+            return text.split("\n\n"), "\n\n"
+        if split_by == "sentence":
+            return text.split("."), "."
+        if split_by == "word":
+            return text.split(" "), " "
+        raise NotImplementedError(
+            "TextDocumentSplitter only supports 'passage', 'sentence' or 'word' split_by options."
+        )
```

Comment:
Not a big deal for this specific component, but by looking at the tests I'm realizing this process loses some characters (namely, the splitting chars). Given that the fix is fairly simple, let's avoid that!

Given [how str.split() works](https://docs.python.org/3/library/stdtypes.html#str.split), is guaranteed to remove the split string after every split except for the last one. So we can safely add back such delimiter to all the elements of the list except the last.

Reply:
I changed the implementation of  `_split_into_units` accordingly and it now adds back the delimiters. I also removed a similar functionality from `_concatenate_units`. It should work now also for `split_length=1`. In future, a cleaning feature just like in the old preprocessor could take care of removing trailing whitespaces, newlines, tabs.

Reply:
The `split_at` can then also become a customizable string or regex.